### ISSUE TITLE
Returning Type[x] and trying to instantiate incorrectly displays inputs.

### ISSUE THREAD

**Issue by prchristie** [microsoft/pylance-release#1303]
## Environment data

-   Language Server version: 2021.5.2
-   OS and version: centOS, 8.3
-   Python version (& distribution if applicable, e.g. Anaconda): python3.8.6

## Expected behaviour

When a function returns `Type[class]`, and the type is attempted to be initialized, the inputs for that type are displayed correctly in the popup. local class methods do not appear when trying to access a member IE `uninstantiated_class.function`

## Actual behaviour

Trying to instantiate the type displays `(*args: Any, **kwds: Any) -> Any` as the type hint. Hovering over the class displays the correct inputs though. Doing a `class_type.` has intellisense for functions that it shouldn't

## Code Snippet / Additional information
Displaying wrong type hint for inputs
![image](https://user-images.githubusercontent.com/60021972/118617276-a0f7bb00-b801-11eb-99f3-ddd64b433c71.png)
Displaying correct type hints when the class is hovered over
![image](https://user-images.githubusercontent.com/60021972/118617344-b0770400-b801-11eb-9478-1e06940d8813.png)
Showing intellisense for local class functions
![image](https://user-images.githubusercontent.com/60021972/118617433-c97fb500-b801-11eb-9f9d-06bba2ccd6c6.png)

This last one is debatable as you can actually access the local functions, its just unlikely that you want them. Feel more than free to just ignore that entirely.

Full code sample:

```python
class Example:
    num = 1

    def __init__(self, input_1: int, input_2: float, input_3: str):
        pass

    def local_function(self):
        pass

    @classmethod
    def cm(cls):
        pass

    @staticmethod
    def sm():
        pass
    
def get_cls()->Type[Example]:
    return Example

example_type = get_cls()

example_type()
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by Azureblade3808** [comment]
Although I agree there is an inconsistency here, I think `(*args: Any, **kwds: Any) -> Any` should be the correct type hint.

The type variable of `Type`/`type` is covariant, so a type conforming to `Type[X]` is not always `X` but can be any subclass of `X` including `X` itself. There is no way to guarantee this type can be instantiated with the same signature of `X`'s initializer.

See my example:
```
from typing import Type

class Base:
    def __init__(self, i: int):
        ...

class Derived(Base):
    def __init__(self, s: str):
        ...

def get_cls() -> Type[Base]:
    return Derived

cls = get_cls()  # Implied type: Type[Base].
_ = cls(i=0)  # No warning from pylance, but met ValueError in runtime.
_ = cls(s="a")  # Runnable, but got a warning from pylance.
```

**Comment by prchristie** [comment]
The example you provided is similar to how I discovered this, and I see where you are coming from. While trying to implement a factory pattern to produce implementations of an interface, I found I couldn't get proper typing. 

So then how do you propose someone would correctly type a factory pattern that returns implementations of an interface? I would personally expect that a child inheriting from an interface doesn't change any of the inherited method inputs, including the initializer, and that if you'd like additional inputs for a child that then an *args or **kwargs can be added to the base class initializer, but like your example shows its more than possible to do such a thing so it makes it hard to generalize this. Maybe I just need to change the pattern being used or there could be some Interface type added that explicitly says inheriting objects will not change the method inputs provided by the given class? I feel like Im missing something obvious here.

**Comment by Azureblade3808** [comment]
You can define a class method to implement a factory pattern.

```python
from __future__ import annotations

class Base:
    @classmethod
    def create(cls, i: int, s: str):
        return Base(i=i)

    def __init__(self, i: int):
        ...

class Derived(Base):
    @classmethod
    def create(cls, i: int, s: str):
        return Derived(s=s)

    def __init__(self, s: str):
        ...

def get_cls() -> type[Base]:
    return Derived

cls = get_cls()  # Implied type: Type[Base]
instance = cls.create(i=0, s="a")  # Implied type: Base

if isinstance(instance, Derived):
    ...
```

The class method `create` is guaranteed to be compatible, otherwise there would be a warning. So calling `create` on a type conforming to `Type[Base]` with the signature of `Base.create(...)` is always legitimate.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.5.3, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202153-19-may-2021
